### DISCUSSION TITLE
Avoid catching all Exceptions in load_model

### DISCUSSION THREAD

**Discussion by rpeng** [mongodb/mongoid#5418] [Ideas]
Would it be possible to change this to allow certain exceptions to re-raise?

https://github.com/mongodb/mongoid/blob/94fdc7d0350eacbfda99710856f5cddbf8c2fc6f/lib/rails/mongoid.rb#L50

Many rails apps pre-load mongo models, and a lot use Zeitwerk to perform the loading. Sometimes, applications will receive SIGTERMs or SIGINTs on startup, and if mongoid happens to be preloading models, this exception would be caught and ignored. This essentially swallows the signal and allows the application to _continue booting_, even though the framework is in an inconsistent state (e.g. half way through pre-loading).

This also can explain why certain rails users are seeing `NameErrors` / uninitialized constants / Zeitwerk load errors.

An example would be like so:
```
    def load_model(file)
      begin
        require_dependency(file)
      rescue SignalException => e
        Logger.new(STDERR).warn("Received signal #{e.message}, raising")
        raise
      rescue Exception => e
        Logger.new(STDERR).warn(e.message)
      end
    end
```

**Comment by edmund-huber** [comment]
https://thoughtbot.com/blog/rescue-standarderror-not-exception

I think mongoid should have a really good reason to be rescuing `Exception` instead of `StandardError`, what was the original rationale for catching `Exception` if any?

**Comment by alexbevi** [comment]
Thanks for the report @rpeng and comment @edmund-huber. I've opened [MONGOID-5463](https://jira.mongodb.org/browse/MONGOID-5463) for our team to triage this issue. Once this has been reviewed further updates will be shared via Jira.
